### PR TITLE
C/C++ wrapper

### PR DESCRIPTION
**Pull request by scarrazza**
Adds an optional C-API to execute QASM circuits.

Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

File: capi/src/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+
+@ffi.def_extern()
+def execute_qasm(circuit, platform, nshots):
+    """Generate samples for a given circuit qasm, platform and number of
+    shots."""
+    py_circuit = ffi.string(circuit).decode("utf-8")
+    py_platform = ffi.string(platform).decode("utf-8")
+    qasm_res = py_execute_qasm(circuit=py_circuit, platform=py_platform, nshots=nshots)
+    return ffi.cast("int*", ffi.from_buffer(qasm_res.samples().ravel()))
```
Comment (stavros11) [inline capi/src/wrapper.py:14]: As @BrunoLiegiBastonLiegi pointed out to me, flattening here may be a bit confusing if more than one qubits are measured, but I guess it is hard to pass the matrix structure to C.

**Review by stavros11 — APPROVED** [review]
Thanks, I tested the example following the README and it works. For people without much experience (eg. me) installation is not very straightforward, but I guess the users this is targeting will know how to do it.

File: capi/src/wrapper.py (hunk 1 above)
Comment (alecandido) [reply capi/src/wrapper.py:14]: This we discussed with @scarrazza, and we were aware of the issue.

However, in C++ there are no matrices in STL, and the vector of vectors (or array of arrays) construction it is not necessarily the most practical one.
There are libraries defining matrices and tensors in C++, but you would have to commit to a single one (with all the issue in C++ dependencies management).

The solution has been to just pass the flattened array, and delegate the reshape in their favorite matrix container to the users.

File: capi/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+SET(CMAKE_SKIP_BUILD_RPATH  FALSE)
+SET(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
+SET(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
+
+project(libqibolab)
+
+set(VERSION "\"0.0.1\"")
+
+find_package(Python3 COMPONENTS Interpreter Development)
```
Comment (alecandido) [inline capi/CMakeLists.txt:11]: For some reason, it is able to find the interpreter, and it doesn't complain
```
-- Found Python3: /Users/alessandro/Projects/Quantum/qibolab/.venv/bin/python3.11 (found version "3.11.7") found components: Interpreter Development Development.Module Development.Embed
generating cqibolab.c
(already up-to-date)
-- Configuring done (3.2s)
-- Generating done (0.0s)
```
but then I'm missing the Python headers, and it crashes during the actual build
<details><summary>Build error</summary>
<p>

```
❯ cmake --build build
[ 50%] Building C object CMakeFiles/qibolab.dir/src/cqibolab.c.o
[100%] Linking C shared library libqibolab.dylib
Undefined symbols for architecture arm64:
  "_PyCapsule_Type", referenced from:
      __cffi_carefully_make_gil in cqibolab.c.o
  "_PyDict_GetItemString", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyDict_New", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyDict_SetItemString", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyErr_Display", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyErr_Fetch", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyErr_NormalizeException", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyErr_Occurred", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyEval_EvalCode", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyEval_GetBuiltins", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyEval_SaveThread", referenced from:
      __cffi_carefully_make_gil in cqibolab.c.o
  "_PyFile_WriteObject", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyFile_WriteString", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyGILState_Ensure", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyGILState_Release", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyImport_GetModuleDict", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyImport_ImportModule", referenced from:
      __cffi_init in cqibolab.c.o
  "_PyLong_FromVoidPtr", referenced from:
      __cffi_init in cqibolab.c.o
  "_PyObject_CallMethod", referenced from:
      __cffi_init in cqibolab.c.o
  "_PyObject_Free", referenced from:
      __cffi_free_array_arguments in cqibolab.c.o
  "_PyObject_GetAttrString", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_PyObject_Malloc", referenced from:
      __cffi_convert_array_argument in cqibolab.c.o
  "_PySys_GetObject", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_Py_CompileStringExFlags", referenced from:
      __cffi_initialize_python in cqibolab.c.o
  "_Py_InitializeEx", referenced from:
      __cffi_py_initialize in cqibolab.c.o
  "_Py_IsInitialized", referenced from:
      __cffi_carefully_make_gil in cqibolab.c.o
  "__Py_Dealloc", referenced from:
      _Py_DECREF in cqibolab.c.o
  "__Py_NoneStruct", referenced from:
      __cffi_initialize_python in cqibolab.c.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/qibolab.dir/build.make:97: libqibolab.dylib] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/qibolab.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

</p>
</details>

(of course I'm not using Ubuntu, but Nix... for which I'm trying to fix by myself...)

File: capi/src/wrapper.py (hunk 1 above)
Comment (scarrazza) [reply capi/src/wrapper.py:14]: Indeed, we could go for more complex solutions, e.g. pybind11, but probably later, if someone really uses this.

File: capi/CMakeLists.txt (hunk 2 above)
Comment (scarrazza) [reply capi/CMakeLists.txt:11]: Should be fixed now.

Comment (alecandido) [reply capi/CMakeLists.txt:11]: I was planning to make the workflow myself, but I first I wanted to compile it (on MacOS, with Nix... it's not being the easiest...)

Comment (scarrazza) [reply capi/CMakeLists.txt:11]: Ok, sounds reasonable.

**Comment by alecandido** [comment]
@scarrazza the workflow is working on Ubuntu. I wanted to try also on MacOS and Windows, but it on Windows (at least) is failing.

The reason is trivial: it seems like the `activate` script is generated at a different path inside the virtual environment (i.e. `env/Scripts/activate`, instead of the usual `env/bin/activate`).
I can easily fix that, but it would (slightly) complicate the workflow...

My question is: do we care at all about running on Windows?

**Comment by scarrazza** [comment]
Thanks, we can drop windows for this test.

**Review by alecandido — APPROVED** [review]
Ok, it would be better for MacOS to work (even in the workflow, but also on Apple Silicon). But it doesn't seem to be trivial, nor urgent.

For the time being, I simply dropped from the workflow.

**Pull request merged**